src/hotspot/share/cds/heapShared.cpp

```diff
@@ -311,6 +314,36 @@ bool HeapShared::archive_object(oop obj) {
     archived_object_cache()->maybe_grow();
     mark_native_pointers(obj);
 
+    Klass* k = obj->klass();
+    if (k->is_instance_klass()) {
+      // Whenever we see a non-array Java object of type X, we mark X to be aot-initialized.
+      // This ensures that during the production run, whenever Java code seens a cached object
+      // of type X, we know that X is already initialized. (see TODO comment below ...)
+
+      if (InstanceKlass::cast(k)->is_enum_subclass()) {
+        // We can't rerun <clinit> of enum classes (see cdsEnumKlass.cpp) so
+        // we must store them as AOT-initialized.
+        AOTArtifactFinder::add_aot_inited_class(InstanceKlass::cast(k));
+      } else if (subgraph_info == _dump_time_special_subgraph) {
```

Comment:
Can this be an "or" condition rather than "else if" since the same function is called for either on of the conditions?

Reply:
Fixed.